src/include/txn.i

```diff
@@ -343,51 +343,58 @@ __wt_txn_unmodify(WT_SESSION_IMPL *session)
 }
 
 /*
- * __wt_txn_op_commit_page_del --
- *	Make the transaction ID and timestamp updates necessary to a ref that
- *      was created by a fast delete truncate operation.
+ * __wt_txn_op_apply_prepare_state --
+ *	Apply the correct prepare state and the timestamp to the ref and to any
+ *      updates in the page del update list.
  */
 static inline void
-__wt_txn_op_commit_page_del(WT_SESSION_IMPL *session, WT_REF *ref)
+__wt_txn_op_apply_prepare_state(
+    WT_SESSION_IMPL *session, WT_REF *ref, bool commit)
 {
 	WT_TXN *txn;
 	WT_UPDATE **updp;
+	wt_timestamp_t ts;
 	uint32_t previous_state;
+	uint8_t prepare_state;
 
 	txn = &session->txn;
 
-	/* Avoid locking the page if a previous eviction already cleaned up. */
-	if (ref->page_del->update_list == NULL)
-		return;
-
 	/*
 	 * Lock the ref to ensure we don't race with eviction freeing the
-	 * page deleted update list.
+	 * page deleted update list or with a page instantiate.
 	 */
 	for (;; __wt_yield()) {
 		previous_state = ref->state;
+		WT_ASSERT(session, previous_state != WT_REF_READING);
 		if (previous_state != WT_REF_LOCKED &&
+		    (previous_state == WT_REF_MEM ||
+		    previous_state == WT_REF_DELETED) &&
 		    __wt_atomic_casv32(
 		    &ref->state, previous_state, WT_REF_LOCKED))
 			break;
 	}
 
+	if (commit) {
+		ts = txn->commit_timestamp;
+		prepare_state = WT_PREPARE_RESOLVED;
+	} else {
+		ts = txn->prepare_timestamp;
+		prepare_state = WT_PREPARE_INPROGRESS;
+	}
 	for (updp = ref->page_del->update_list;
 	    updp != NULL && *updp != NULL; ++updp) {
-		(*updp)->timestamp = txn->commit_timestamp;
-		if (F_ISSET(txn, WT_TXN_PREPARE))
```

Comment:
Was this just redundant before? Should we be asserting it now?

Reply:
I think this was redundant before, this function was called in an if condition that already checked for WT_TXN_PREPARE state. Now this function can be called as part of prepare() call, before WT_TXN_PREPARE is set on txn, so we shouldn't be asserting. Now, this function will be called with and without WT_TXN_PREPARE set on txn.